evalml/automl/progress.py

```diff
@@ -0,0 +1,172 @@
+
+        if self.max_time and elapsed >= self.max_time:
+            return False
+        elif self.max_iterations and self.current_iterations >= self.max_iterations:
+            return False
+        elif (
+            self.max_batches
+            and self.current_batches >= self.max_batches
+            and not mid_batch
+        ):
+            return False
+
+        # check for early stopping
+        if self.patience is None or self.tolerance is None:
+            return True
+
+        first_id = results["search_order"][0]
+        best_score = results["pipeline_results"][first_id]["mean_cv_score"]
+        num_without_improvement = 0
+        for id in results["search_order"][1:]:
+            curr_score = results["pipeline_results"][id]["mean_cv_score"]
+            significant_change = (
+                abs((curr_score - best_score) / best_score) > self.tolerance
+            )
+            score_improved = (
+                curr_score > best_score
+                if self.objective.greater_is_better
+                else curr_score < best_score
+            )
+            if score_improved and significant_change:
+                best_score = curr_score
+                num_without_improvement = 0
+            else:
+                num_without_improvement += 1
+            if num_without_improvement >= self.patience:
+                self.logger.info(
+                    "\n\n{} iterations without improvement. Stopping search early...".format(
+                        self.patience,
+                    ),
+                )
+                return False
+        return True
+
+    def _build_progress_dict(self, stopping_criteria, current_state, end_state, unit):
+        progress_dict = {}
+        progress_dict["stopping_criteria"] = stopping_criteria
+        progress_dict["current_state"] = current_state
+        progress_dict["end_state"] = end_state
+        progress_dict["unit"] = unit
+        return progress_dict
+
+    def return_progress(self):
+        """Return information about current and end state of each stopping criteria in order of priority.
+
+        Returns:
+            List[Dict[str, unit]]: list of dictionaries containing information of each stopping criteria.
+        """
+        progress = []
+        if self.max_time:
+            progress.append(
+                self._build_progress_dict(
+                    "max_time",
+                    self.current_time - self._start_time,
+                    self.max_time,
+                    "seconds",
+                ),
+            )
+        if self.max_iterations or self.max_batches:
+            max_iterations = (
+                self.max_iterations
+                if self.max_iterations
+                else sum(
+                    [
+                        self.automl_algorithm.num_pipelines_per_batch(n)
+                        for n in range(self.max_batches)
+                    ],
+                )
+            )
```

Comment:
I'm a little confused by the logic here - if `max_batches` is the stopping criteria, we add stopping conditions for both the number of batches and the number of iterations, but we calculate the number of iterations based on the number of batches. Isn't that redundant/ is there a reason we need both in this case?

I see in the tests that you have an example of default algorithm setting both `max_iterations` and `max_batches`, but is that possible when using AutoMLSearch outside of testing?

Reply:
I don't believe we add `max_iterations` as stopping criteria in search (I might be confusing your point here). The reason for this conversion is to allow users to ping the number of pipelines progress/left if they only select `max_batches` as stopping criteria. I can see how that is confusing though. I wanted to keep the structure the same if a user did select both `max_iterations` and `max_batches` (`max_iterations` would take precedence) but maybe I could add the pipeline information to the `max_batches` dictionary instead?